torch/_subclasses/functional_tensor.py

```diff
@@ -220,6 +220,21 @@ def __init__(self, pre_dispatch=False, export=False):
         self._mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
         # This will be turned off later for pre-dispatch functionalization
         self._dispatch_key = torch._C.DispatchKey.PreDispatch if pre_dispatch else None  # type: ignore[attr-defined]
+        # Map of effect type (ex. _EffectType.ORDERED) to a token. The tokens help keep
+        # track of the ordering between side effectful operations.
+        self._tokens: Dict[Any, torch.Tensor] = {}
+
+        # Functionalization runs twice in AOTAutograd, once in
+        # `run_functionalized_fw_and_collect_metadata` to collect metadata to
+        # see which tensors need to be functionalized and how many tokens we
+        # need, and another time in `make_fx` which does the actual tracing to
+        # replace ops with their functional variants and handling side-effectful
+        # ops. We need to distinguish these two tracing modes for handling
+        # side-effectful ops. In the first functionalization pass, where
+        # tracing=False, we will count the number of tokens needed. In the
+        # second functionalization pass, when tracing=True, we will insert
+        # tokens into the graph.
+        self._tracing = _tracing
```

Comment:
To clarify: we just need this flag to have a nice assertion, right?

Reply:
If we keep this, could we make the name a bit more descriptive? If someone just wants to use FunctionalTensorMode outside of AOTAutograd, it's not clear if they should set _tracing to be True or False. What the flag is really saying is something like "hey, you've got tokens, use these tokens"

Reply:
Hmm I'm pretty neutral on the new argument. Alternatively, you could avoid adding the new state by querying whether there is currently a proxy mode active (using `torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.PROXY)`). This avoids the state, although it is definitely more less explicit (we're implicitly relying on the fact that aot autograd runs functionalization twice, once without proxy mode and once with proxy mode. Although this is pretty unlikely to change).